src/main/java/de/tum/in/www1/artemis/service/OrganizationService.java

```diff
@@ -0,0 +1,124 @@
+
+    private final OrganizationRepository organizationRepository;
+
+    private final UserRepository userRepository;
+
+    private final CourseRepository courseRepository;
+
+    public OrganizationService(OrganizationRepository organizationRepository, UserRepository userRepository, CourseRepository courseRepository) {
+        this.organizationRepository = organizationRepository;
+        this.userRepository = userRepository;
+        this.courseRepository = courseRepository;
+    }
+
+    /**
+     * Performs indexing over all users using the email pattern of the provided organization.
+     * Users matching the pattern will be added to the organization.
+     * Users not matching the pattern will be removed from the organization (if contained).
+     * @param organization the organization used to perform the indexing
+     */
+    public void indexing(final Organization organization) {
+        log.debug("Start indexing for organization: {}", organization.getName());
+        organization.getUsers().forEach(user -> {
+            userRepository.removeOrganizationFromUser(user.getId(), organization);
+        });
+        userRepository.findAllMatchingEmailPattern(organization.getEmailPattern()).forEach(user -> {
+            log.debug("User {} matches {} email pattern. Adding", user.getLogin(), organization.getName());
+            userRepository.addOrganizationToUser(user.getId(), organization);
+        });
+    }
+
+    /**
+     * Save an organization
+     *
+     * @param organization the entity to save
+     * @return the persisted entity
+     */
+    public Organization save(Organization organization) {
+        log.debug("Request to save Organization : {}", organization.getName());
+        return organizationRepository.save(organization);
+    }
+
+    /**
+     * Add a new organization and execute indexing based on its emailPattern
+     * @param organization the organization to add
+     * @return the persisted organization entity
+     */
+    public Organization add(Organization organization) {
+        Organization addedOrganization = save(organization);
+        addedOrganization = organizationRepository.findOneWithEagerUsersAndCoursesOrElseThrow(addedOrganization.getId());
+        indexing(addedOrganization);
+        return addedOrganization;
+    }
+
+    /**
+     * Update an organization
+     * To avoid removing the currently mapped users and courses of the organization,
+     * these are loaded eagerly and the edited values changed within the loaded entity.
+     * @param organization the organization to update
+     * @return the updated organization
+     */
+    public Organization update(Organization organization) {
+        log.debug("Request to update Organization : {}", organization);
+        boolean indexingRequired = false;
+        Optional<Organization> optionalOldOrganization = organizationRepository.findByIdWithEagerUsersAndCourses(organization.getId());
```

Comment:
```suggestion
        var oldOrganization = organizationRepository.findByIdWithEagerUsersAndCoursesElseThrow(organization.getId());
```

Reply:
Added in https://github.com/ls1intum/Artemis/pull/3102